rmw_zenoh_cpp/src/detail/graph_cache.cpp

```diff
@@ -41,170 +41,284 @@ using Entity = liveliness::Entity;
-            entity_desc = "service";
-            return graph_node.services_;
+  // Then update the variables tracking topics across the graph.
+  // We invoke update_topic_map_for_put() with report_events set to true for
+  // pub/sub.
+  if (entity.type() == EntityType::Publisher ||
+    entity.type() == EntityType::Subscription)
+  {
+    update_topic_map_for_put(this->graph_topics_, entity, true);
+  } else {
+    update_topic_map_for_put(this->graph_services_, entity);
+  }
+}
 
+///=============================================================================
+void GraphCache::update_topic_map_for_put(
+  GraphNode::TopicMap & topic_map,
+  const liveliness::Entity & entity,
+  bool report_events)
+{
+  if (!entity.topic_info().has_value()) {
+    // This should not happen as topic_info should be populated for all non-node entites.
+    RCUTILS_LOG_WARN_NAMED(
+      "rmw_zenoh_cpp",
+      "update_topic_map_for_put() called for non-node entity without valid TopicInfo. "
+      "Report this.");
+    return;
+  }
+  const liveliness::TopicInfo topic_info = entity.topic_info().value();
+
+  // For the sake of reusing data structures and lookup functions, we treat publishers and
+  // clients as equivalent. Similarly, subscriptions and services are equivalent.
+  const bool is_pub = is_entity_pub(entity);
+  // Initialize a map that will be populated with any QoS events that may be detected.
+  std::unordered_map<liveliness::Entity, std::unordered_set<rmw_zenoh_event_type_t>>
+  local_entities_with_events = {};
+  std::unordered_set<liveliness::Entity> pubs = {};
+  std::unordered_set<liveliness::Entity> subs = {};
+  if (is_pub) {
+    pubs.insert(entity);
+  } else {
+    subs.insert(entity);
+  }
+  TopicDataPtr graph_topic_data = std::make_shared<TopicData>(
+    topic_info,
+    std::move(pubs),
+    std::move(subs));
+
+  std::string qos_str = liveliness::qos_to_keyexpr(topic_info.qos_);
+  GraphNode::TopicQoSMap topic_qos_map = {
+    {qos_str, graph_topic_data}};
+
+  GraphNode::TopicMap::iterator topic_map_it = topic_map.find(topic_info.name_);
+  if (topic_map_it == topic_map.end()) {
+    // First time this topic name is discovered for the node so we insert a TopicTypeMap.
+    GraphNode::TopicTypeMap topic_data_map = {
+      {topic_info.type_, std::move(topic_qos_map)}
+    };
+    topic_map.insert(std::make_pair(topic_info.name_, std::move(topic_data_map)));
+  } else {
+    // The topic exists for the node so we check if the type also exists.
+    GraphNode::TopicTypeMap::iterator topic_type_map_it = topic_map_it->second.find(
+      topic_info.type_);
+    if (topic_type_map_it == topic_map_it->second.end()) {
+      // First time this topic type is added.
+      // TODO(Yadunund) Check for and report an *_INCOMPATIBLE_TYPE events.
+
+      topic_map_it->second.insert(
+        std::make_pair(
+          topic_info.type_,
+          std::move(topic_qos_map)));
+    } else {
```

Comment:
This entire function is pretty large and deeply nested.  Is there a way to refactor this to avoid that somehow?  Split this into functions, maybe?

Reply:
Agreed. I've simplified things here https://github.com/ros2/rmw_zenoh/pull/103/commits/47973be62df1ea92f5387f9a7c5d5d5caec1e227